src/nvim/mark_extended.c

```diff
@@ -0,0 +1,907 @@
+                           buf_T *buf,
+                           uint64_t ns,
+                           uint64_t id,
+                           linenr_T lnum,
+                           colnr_T col,
+                           ExtmarkReverseType undo)
+{
+  if (undo != extmarkNoUndo) {
+    u_extmark_update(buf, ns, id, extmark->line->lnum, extmark->col,
+                     lnum, col);
+  }
+  extmark->col = col;
+  extmark->line->lnum = lnum;
+}
+
+static int extmark_delete(ExtendedMark *extmark,
+                          buf_T *buf,
+                          uint64_t ns,
+                          uint64_t id,
+                          ExtmarkReverseType undo)
+{
+  if (undo != extmarkNoUndo) {
+    u_extmark_set(buf, ns, id, extmark->line->lnum, extmark->col, kExtmarkUnset);
+  }
+
+  // Remove our key from the namespace
+  ExtmarkNs *ns_obj = pmap_get(uint64_t)(buf->b_extmark_ns, ns);
+  pmap_del(uint64_t)(ns_obj->map, id);
+
+  // Remove the mark mark from the line
+  ExtMarkLine *extline = extmark->line;
+  kb_del(markitems, &extline->items, *extmark);
+  // Remove the line if there are no more marks in the line
+  if (kb_size(&extline->items) == 0) {
+    kb_del(extlines, &buf->b_extlines, extline);
+  }
+  return true;
+}
+
+ExtendedMark *extmark_from_id(buf_T *buf, uint64_t ns, uint64_t id)
+{
+  if (!buf->b_extmark_ns) {
+    return NULL;
+  }
+  ExtmarkNs *ns_obj = pmap_get(uint64_t)(buf->b_extmark_ns, ns);
+  if (!ns_obj || !kh_size(ns_obj->map->table)) {
+    return NULL;
+  }
+  ExtMarkLine *extline = pmap_get(uint64_t)(ns_obj->map, id);
+  if (!extline) {
+    return NULL;
+  }
+
+  FOR_ALL_EXTMARKS_IN_LINE(&extline->items, {
+    if (extmark->ns_id == ns
+        && extmark->mark_id == id) {
+      return extmark;
+    }
+  })
+  return NULL;
+}
+
+ExtendedMark *extmark_from_pos(buf_T *buf,
+                               uint64_t ns, linenr_T lnum, colnr_T col)
+{
+  if (!buf->b_extmark_ns) {
+    return NULL;
+  }
+  FOR_ALL_EXTMARKS(buf, ns, lnum, col, lnum, col, {
+    if (extmark->ns_id == ns) {
+      if (extmark->col == col) {
+        return extmark;
+      }
+    }
+  })
+  return NULL;
+}
+
+void extmark_free_all(buf_T *buf)
+{
+  if (!buf->b_extmark_ns) {
+    return;
+  }
+  ExtmarkNs *ns_obj;
+  map_foreach_value(buf->b_extmark_ns, ns_obj, {
+    pmap_free(uint64_t)(ns_obj->map);
+    xfree(ns_obj);
+  });
+  pmap_free(uint64_t)(buf->b_extmark_ns);
+
+  FOR_ALL_EXTMARKLINES(buf, 0, -1, {
+    xfree(extline);
+  })
+  kb_destroy(extlines, (&buf->b_extlines));
+
+  // This is emptied after a move
+  kv_destroy(buf->b_extmark_end_temp_space);
+}
+
+// TODO(timeyyy): make this non static ..
+static u_header_T *get_undo_header(buf_T *buf)
+{
+  u_header_T *uhp = NULL;
+  if (buf->b_u_curhead != NULL) {
+    uhp = buf->b_u_curhead;
+  } else if (buf->b_u_newhead) {
+    uhp = buf->b_u_newhead;
+  }
+  if (!uhp) {
+    // TODO(timeyyy): create the first undo header...
```

Comment:
I can't find a way to create the undo header if it doesn't exist. Currently any extmarks stuff will fail unless an undo header has been created (e.g entering text in insert mode).